### PR TITLE
Highlight linked comment

### PR DESCRIPTION
**Pull request by perunt**
### Details
This PR is part of the Comment Linking project. Specifically, this task allows users to see which message they referred to by changing the background color of the linked message. WARNING: it doesn't scroll to that message; it merely changes the background color. This background color remains until the user navigates back, so it will be there permanently as long as it remains in the route.

### Fixed Issues

$ https://github.com/Expensify/App/issues/23230
PROPOSAL: 

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps
Enable the beta flag named 'commentLinking'.
- Open a chat.
- Copy the link to the desired message.
- Navigate to another chat.
- Paste the copied link into this chat.
- Click on the pasted link. You were navigated to the separate chat
- Scroll to the linked message if it's not immediately visible.
- Verify that the background color of the linked message has been changed.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/48593211/a531187d-dae9-4251-a90e-def22813f447

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/48593211/e819f542-37a5-49bc-847c-a1da319b1698

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/48593211/359fc68b-16c7-4e44-a2e3-4198461ff5d4

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/48593211/2b2ae95a-efd9-47ba-a516-78113b0c244e

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/48593211/5c33e251-1ba4-4792-a7c8-e45b04e623ec

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/48593211/e7be4062-2441-42c0-9599-09c2ae54bab9

</details>

### CODE REVIEW

**Comment by perunt** [comment]
@situchan, could you take a look at this? It's a part of the comment-linking feature, and it is a big priority.

**Comment by perunt** [comment]
@situchan, could you take a look at this? It's a part of the comment-linking feature, and it is a big priority. Thanks!

**Comment by situchan** [comment]
reviewing now. can you please fix conflict?

**Comment by perunt** [comment]
@situchan can you take a look at this?

File: src/pages/home/report/ReportActionItem.js [hunk 1]
```diff
@@ -687,6 +695,7 @@ export default compose(
             key: ({action}) => `${ONYXKEYS.COLLECTION.REPORT_ACTIONS_REACTIONS}${action.reportActionID}`,
         },
     }),
+    withRoute,
```
Comment (situchan) [inline src/pages/home/report/ReportActionItem.js:698]: As `ReportActionItem` is function component, why cannot we use `useRoute` hook directly?

Comment (perunt) [reply src/pages/home/report/ReportActionItem.js:698]: Yeah, you're right. We are moving away from HOCs and transitioning to functional components. Thanks for pointing it out. I'll update it shortly

**Comment by perunt** [comment]
@situchan is it good for you now?

**Comment by situchan** [comment]
yes good, will checklist in a few hrs

**Comment by situchan** [comment]
Server is down - https://expensify.slack.com/archives/C01GTK53T8Q/p1695314249132269
Will checklist once it's back

**Comment by perunt** [comment]
Sure, thanks for keeping it up to date!

**Comment by perunt** [comment]
@situchan, just a friendly reminder. As we need to deliver the whole that features by the end of the month

**Comment by situchan** [comment]
Thanks for the reminder. Reviewing now

File: src/pages/home/report/ReportActionItem.js [hunk 2]
```diff
@@ -138,6 +141,12 @@ function ReportActionItem(props) {
     const prevDraftMessage = usePrevious(props.draftMessage);
     const originalReportID = ReportUtils.getOriginalReportID(props.report.reportID, props.action);
     const originalReport = props.report.reportID === originalReportID ? props.report : ReportUtils.getReport(originalReportID);
+    const reportActionID = lodashGet(route, 'params.reportActionID');
+
+    const highlightedBackgroundColorIfNeeded = useMemo(
+        () => (reportActionID === props.action.reportActionID ? {backgroundColor: themeColors.highlightBG} : {}),
```
Comment (situchan) [inline src/pages/home/report/ReportActionItem.js:147]: We don't prefer inline style.
```suggestion
        () => (reportActionID === props.action.reportActionID ? StyleUtils.getBackgroundColorStyle(themeColors.highlightBG) : {}),
```

**Comment by situchan** [comment]
Please pull main

File: src/pages/home/report/ReportActionItem.js (hunk 2 above)
Comment (perunt) [reply src/pages/home/report/ReportActionItem.js]: 👍🏻

**Comment by situchan** [comment]
Once I click message deep link, no way to remove highlight unless browser url is changed.
Is this out of scope for this PR?

https://github.com/Expensify/App/assets/108292595/fbb3de4a-a69b-4cc5-af5e-00ebbb941a73

**Comment by situchan** [comment]
It's odd to me that money request preview is not highlighted while message is highlighted

Shouldn't this be consistent with hover effect?

Screenshot 2023-09-22 at 9 40 57 PM

Screenshot 2023-09-22 at 9 41 04 PM

**Comment by situchan** [comment]
Given this is priority project, I'd like to consider above concerns out of scope here

**Comment by situchan** [comment]
It would be good to change message deep link back to report link when click active report in LHN

i.e. `/r/8594040674715855/3686280274134926037` to `/r/8594040674715855`

https://github.com/Expensify/App/assets/108292595/53a38372-80e4-4671-855e-4b3b2707c387

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



https://github.com/Expensify/App/assets/108292595/a89b9bd1-fe16-4fa8-bc98-3c358a9eb750


</details>

<details>
<summary>Mobile Web - Chrome</summary>


https://github.com/Expensify/App/assets/108292595/7d5d6252-3dbf-4d60-9bc9-5097133d6c18



</details>

<details>
<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/108292595/ce1650f8-d3db-46ce-8cc7-5a6ed0e20855



</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/108292595/43c91127-d188-448c-9b36-a92862e72c41



</details>

<details>
<summary>iOS</summary>



https://github.com/Expensify/App/assets/108292595/dc8df589-e2fb-475b-9440-1ec413c698de


</details>

<details>
<summary>Android</summary>



https://github.com/Expensify/App/assets/108292595/04c77d6e-3106-45be-9bd1-963be906e69a



</details>

**Review by situchan — APPROVED** [review]
:shipit: 

Known issues/concerns that can be follow-up:
https://github.com/Expensify/App/pull/27227#issuecomment-1731633504
https://github.com/Expensify/App/pull/27227#issuecomment-1731637411
https://github.com/Expensify/App/pull/27227#issuecomment-1731646394

**Review by situchan — CHANGES_REQUESTED** [review]
Seems you're addressing [feedback](https://github.com/Expensify/App/pull/27227#pullrequestreview-1640353209) here

**Comment by perunt** [comment]
@situchan Yes, I have been addressing those comments. Also, I had a weird test error that was not related to this PR. Now it should be good.

**Comment by Gonals** [comment]
All good here? @situchan, do we need to retest?

**Comment by situchan** [comment]
I will review shortly

**Comment by Gonals** [comment]
> I will review shortly

Sounds good. Let me know when you are done, @situchan

**Comment by Gonals** [comment]
@perunt, seems like we now have some conflicts

**Comment by perunt** [comment]
@Gonals done

File: src/pages/home/report/ReportActionItem.js [hunk 3]
```diff
@@ -127,6 +129,7 @@ const defaultProps = {
 };
 
 function ReportActionItem(props) {
+    const route = useRoute();
```
Comment (situchan) [inline src/pages/home/report/ReportActionItem.js:132]: For performance, I think it's better `useRoute` only once in parent view (report screen.)
And pass only flag to `ReportActionItem`.
It's not performant to subscribe route change on every item.

Comment (Gonals) [reply src/pages/home/report/ReportActionItem.js:132]: @perunt ⬆️

Comment (perunt) [reply src/pages/home/report/ReportActionItem.js:132]: In that case, we will rerender all messages if we want to highlight a message already on the screen.
I'll make changes in a moment.

Comment (perunt) [reply src/pages/home/report/ReportActionItem.js]: done

**Comment by perunt** [comment]
@Gonals @situchan, is it good now?

**Comment by situchan** [comment]
Some known issues:

- double slash on copy from native
i.e. https://new.expensify.com//r/1273261174467626/1467080164206627159

- composer is blurred when click same chat on LHN


https://github.com/Expensify/App/assets/108292595/c0759fca-804b-49c5-a9e8-f2843b3c5417

**Review by situchan — APPROVED** [review]

**Review by Gonals — APPROVED** [review]

**Comment by kavimuru** [comment]
@Gonals @situchan How do we enable 

>  beta flag named 'commentLinking'.

**Comment by situchan** [comment]
> @Gonals @situchan How do we enable
> 
> > beta flag named 'commentLinking'.

I am not sure. Thought QA team's test accounts have all beta enabled.

@kavimuru are you able see this option in your account?

Screenshot 2023-10-03 at 12 18 58 AM

**Comment by kavimuru** [comment]
@situchan yes we do. Thanks

**Pull request merged**